Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -1051,18 +1051,19 @@ void MatrixWorkspace::setDistribution(bool newValue) {
  *  Whether the workspace contains histogram data
  *  @return whether the workspace contains histogram data
  */
-bool MatrixWorkspace::isHistogramData() const {
+bool MatrixWorkspace::isHistogramData(const std::size_t index) const {
```

Comment:
Nice explanation of the issue and I see what is happening now.

The only thing I'm slightly uncomfortable with here is that as this is part of the public API it suggests that it might be possible for each index to return a different answer to this question when in fact we use in quite a few places the fact that the whole workspace is in either point/histogram mode.

Would we be able to leave the original `bool MatrixWorkspace::isHistogramData() const` method not taking an index but maybe have a private method that accepts an index so that this can be used in `yIndexOfX`? The public method can then just call this with `index==0`?

Reply:
It looks to me as though it is possible to create a workspace with histograms having different xmodes using the MatrixWorkspace::setHistogram method so on that basis I was thinking that changing the API was a positive thing. Perhaps this is a loophole though that we don't want to advertise or encourage?

Fine to make the changes but thought I'd double check first

Reply:
You're right it's a fair point that those set methods do allow you to set arbitrary histograms without checking but those are fairly low level and if we do allow/encourage it I'm a little worried that other places are not set up to cope with it properly. Could preserve the current behaviour for now?

Reply:
I'll record one comment against this in case there are any more problems in future with the multithreaded section of this algorithm. There's a comment against the function cow_ptr<DataType>::access() which says that the function isn't threadsafe. I wonder if this is what I came across in this issue since the X\Y\E vectors in the histograms are managed as cow ptrs. 

All the threads were only writing to a single histogram even before this PR so it seems like there shouldn't have been a problem. Perhaps the copy on write functionality isn't thread safe? One option might be to lock the mutex higher up outside of the if statement but that may have a negative performance impact

```
template <typename DataType> DataType &cow_ptr<DataType>::access() {
  // Use a double-check for sharing so that we only acquire the lock if
  // absolutely necessary
  if (!Data.unique()) {
    std::lock_guard<std::mutex> lock{copyMutex};
    // Check again because another thread may have taken copy and dropped
    // reference count since previous check
    if (!Data.unique()) {
      std::atomic_store(&Data, std::make_shared<DataType>(*Data));
    }
  }
  return *Data;
}
```